packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,220 @@
+          } else {
+            // If syncBlockByRoot could not find anchorBlock, loop and try again with another peer
+            continue;
+          }
+        }
+
+        const oldestSlotRequired = this.getOldestRequiredSlot();
+        const anchorBlockSlot = this.anchorBlock.message.slot;
+        if (anchorBlockSlot <= oldestSlotRequired) {
+          // Sync completed, return the oldestSlotSynced
+          return anchorBlockSlot;
+        }
+
+        const toSlot = anchorBlockSlot;
+        const fromSlot = Math.max(toSlot - this.opts.batchSize, oldestSlotRequired);
+        this.logger.debug("BackfillSync syncing range", {fromSlot, toSlot});
+        await this.syncRange(peer, fromSlot, toSlot, this.anchorBlock.message.parentRoot);
+      } catch (e) {
+        this.logger.debug("BackfillSync sync error", e);
+
+        if (e instanceof BackfillSyncError) {
+          switch (e.type.code) {
+            case BackfillSyncErrorCode.INVALID_SIGNATURE:
+            case BackfillSyncErrorCode.NOT_LINEAR:
+            case BackfillSyncErrorCode.NOT_ANCHORED:
+              this.network.reportPeer(peer, PeerAction.LowToleranceError, "BadSyncBlocks");
+          }
+        }
+      } finally {
+        this.processor.trigger();
+      }
+    }
+
+    throw new ErrorAborted("BackfillSync");
+  }
+
+  private getOldestRequiredSlot(): number {
+    const oldestRequiredEpoch = Math.max(
+      GENESIS_EPOCH,
+      this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+    );
+    return computeStartSlotAtEpoch(oldestRequiredEpoch);
+  }
+
+  private addPeer = (peerId: PeerId, peerStatus: phase0.Status): void => {
+    const requiredSlot =
+      this.anchorBlock?.message.slot ?? computeStartSlotAtEpoch(this.chain.getFinalizedCheckpoint().epoch);
+    this.logger.debug("BackfillSync add peer", {peerhead: peerStatus.headSlot, requiredSlot});
+    if (peerStatus.headSlot >= requiredSlot) {
+      this.peers.add(peerId);
+      this.processor.trigger();
+    }
+  };
+
+  private removePeer(peerId: PeerId): void {
+    this.peers.delete(peerId);
+  }
+
+  private async syncBlockByRoot(peer: PeerId, root: Root): Promise<void> {
+    const [block] = await this.network.reqResp.beaconBlocksByRoot(peer, [root] as List<Root>);
+    if (block) {
+      await verifyBlockProposerSignature(this.chain.bls, this.chain.getHeadState(), [block]);
+
+      await this.db.blockArchive.put(block.message.slot, block);
+      this.anchorBlock = block;
+    }
+  }
+
+  private async syncRange(peer: PeerId, from: Slot, to: Slot, anchorRoot: Root): Promise<void> {
+    const blocks = await this.network.reqResp.beaconBlocksByRange(peer, {startSlot: from, count: to - from, step: 1});
+    if (blocks.length === 0) {
+      return;
+    }
+
+    verifyBlockSequence(this.config, blocks, anchorRoot);
+    await verifyBlockProposerSignature(this.chain.bls, this.chain.getHeadState(), blocks);
```

Comment:
Splited the verifyBlock function into two so it can be reused in `syncBlockByRoot`. @mpetrunic Could you please add a unit test for verifyBlockSequence? Should be not complex to test with actual data, this logic is very important.